File: prismpath-telemetry-rs/tests/test_decode.rs
use prismpath_rs::{parse, V};
use prismpath_telemetry_rs::{decode, quantizer, wire};
use std::collections::HashMap;

const CAT: &str = r#"---
name: cat
start: classify
---
## classify
-> urgent: when kind == 'urgent'
-> batch: when kind in ('nightly', 'weekly')
-> blocked: when status != 'ok'
-> normal: else
## urgent
## batch
## blocked
## normal
"#;

#[test]
fn test_decode_reproduces_routes() {
    let graph = parse(CAT);
    let parts = quantizer::build_partitions(&graph);
    let readings = vec![
        {
            let mut reading = HashMap::new();
            reading.insert("kind".to_string(), V::Str("urgent".to_string()));
            reading.insert("status".to_string(), V::Str("ok".to_string()));
            reading
        },
        {
            let mut reading = HashMap::new();
            reading.insert("kind".to_string(), V::Str("nightly".to_string()));
            reading.insert("status".to_string(), V::Str("ok".to_string()));
            reading
        },
        {
            let mut reading = HashMap::new();
            reading.insert("kind".to_string(), V::Str("adhoc".to_string()));
            reading.insert("status".to_string(), V::Str("bad".to_string()));
            reading
        },
        {
            let mut reading = HashMap::new();
            reading.insert("kind".to_string(), V::Str("weekly".to_string()));
            reading.insert("status".to_string(), V::Str("degraded".to_string()));
            reading
        },
    ];
    let bits = decode::encode_readings(&parts, &readings).unwrap();
    let rep = decode::inspect(&graph, &bits);
    assert_eq!(rep.n_readings, 4);
    assert_eq!(rep.trailing_ints, 0);

    for (orig, row) in readings.iter().zip(rep.readings.iter()) {
        assert_eq!(
            row.routes.get("classify").cloned().flatten(),
            wire::route_node(&graph, "classify", orig)
        );
    }
}

#[test]
fn test_other_renders_readably() {
    let graph = parse(CAT);
    let parts = quantizer::build_partitions(&graph);
    let mut reading = HashMap::new();
    reading.insert("kind".to_string(), V::Str("adhoc".to_string()));
    reading.insert("status".to_string(), V::Str("bad".to_string()));
    let bits = decode::encode_readings(&parts, &[reading]).unwrap();
    let rep = decode::inspect(&graph, &bits);
    assert_eq!(
        rep.readings[0].reading.get("kind").unwrap(),
        &V::Str("<other>".to_string())
    );
}

#[test]
fn test_partial_final_frame_is_reported_not_crashed() {
    let graph = parse(CAT);
    let parts = quantizer::build_partitions(&graph);
    let mut reading = HashMap::new();
    reading.insert("kind".to_string(), V::Str("urgent".to_string()));
    reading.insert("status".to_string(), V::Str("ok".to_string()));
    let bits = decode::encode_readings(&parts, &[reading]).unwrap();
    let rep = decode::inspect(&graph, &format!("{}0", bits));
    assert_eq!(rep.n_readings, 1);
}

File: prismpath/telemetry/tests/test_decode.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Crystal Warden Supply Chain Labs LLC
"""The inspect path: decoding a captured bitstream against its flow reproduces each reading's routing
decision (the .md is the decoder), renders categorical 'other' readably, and survives a partial final
frame."""
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
sys.path.insert(0, str(Path(__file__).resolve().parent.parent.parent.parent))   # repo root
from prismpath.telemetry import decode as dec    # noqa: E402
from prismpath.telemetry import quantizer   # noqa: E402
from prismpath.telemetry import wire        # noqa: E402
from prismpath.kernel.parser import parse  # noqa: E402

CAT = """---
name: cat
start: classify
---
## classify
-> urgent: when kind == 'urgent'
-> batch: when kind in ('nightly', 'weekly')
-> blocked: when status != 'ok'
-> normal: else
## urgent
## batch
## blocked
## normal
"""


def test_decode_reproduces_routes():
    graph = parse(CAT)
    parts = quantizer.build_partitions(graph)
    readings = [{"kind": "urgent", "status": "ok"},
                {"kind": "nightly", "status": "ok"},
                {"kind": "adhoc", "status": "bad"},
                {"kind": "weekly", "status": "degraded"}]
    bits = dec.encode_readings(parts, readings)
    rep = dec.inspect(graph, bits)
    assert rep["n_readings"] == 4 and rep["trailing_ints"] == 0
    # each decoded reading routes exactly as the original did
    for orig, row in zip(readings, rep["readings"]):
        assert row["routes"]["classify"] == wire.route_node(graph, "classify", orig)


def test_other_renders_readably():
    graph = parse(CAT)
    parts = quantizer.build_partitions(graph)
    bits = dec.encode_readings(parts, [{"kind": "adhoc", "status": "bad"}])
    rep = dec.inspect(graph, bits)
    # 'adhoc' is not a listed kind -> the reconstructed representative shows as <other>, not a control char
    assert rep["readings"][0]["reading"]["kind"] == "<other>"


def test_partial_final_frame_is_reported_not_crashed():
    graph = parse(CAT)
    parts = quantizer.build_partitions(graph)
    bits = dec.encode_readings(parts, [{"kind": "urgent", "status": "ok"}])
    rep = dec.inspect(graph, bits + "0")           # a dangling bit = incomplete final frame
    assert rep["n_readings"] == 1              # the complete reading still decodes
